flocker/node/agents/test/test_blockdevice.py

```diff
@@ -698,6 +752,186 @@ def add_application_with_volume(node_state):
+        # ``BlockDeviceVolume`` instances that are connected to remote nodes.
+        # This is an opaque identifier, and just needs to be different than
+        # compute_instance_id for the sake of testing this node.
+        fake_instance_id = u"remote_node_{}".format(
+            unicode(remote_node_state.uuid))
+        remote_volumes = [
+            _create_test_blockdevice_volume_for_dataset_id(
+                dataset_id, attached_to=fake_instance_id)
+            for dataset_id in (remote_node_state.manifestations or {}).keys()]
+
+        volumes.extend(remote_volumes)
+    return volumes
+
+
+# Constants for determining volumes in calculate_changes tests. See
+# _create_block_device_deployer_local_state for details.
+_GET_VOLUMES_FROM_API = NamedConstant()
+_INFER_VOLUMES_FROM_STATE = NamedConstant()
+
+
+def _create_block_device_deployer_local_state(block_device_deployer,
+                                              cluster_state,
+                                              volumes=_GET_VOLUMES_FROM_API,
+                                              compute_instance_id=None):
+    """
+    Creates a BlockDeviceDeployerLocalState from the given cluster_state and
+    optional volumes.
+
+    :param BlockDeviceDeployer block_device_deployer: The
+        ``BlockDeviceDeployer`` instance to get the local state for.
+
+    :param DeploymentState cluster_state: The current state of the cluster.
+
+    :param volumes: A sequence of ``BlockDeviceVolume`` instances representing
+        all volumes in the cluster, _INFER_VOLUMES_FROM_STATE, or
+        _GET_VOLUMES_FROM_API. _GET_VOLUMES_FROM_API is the default, and gets
+        the volumes from the ``IBlockDeviceAPI`` that the block_device_deployer
+        is configured to use. _INFER_VOLUMES_FROM_STATE assumes that the
+        cluster_state was not set up using the API, and thus inspects
+        cluster_state to determine which volumes exist in the cluster. Note
+        that the blockdevice_ids of the volumes that are constructed in this
+        manner are not compatible with the ``IBlockDeviceAPI``.
+
+    :param unicode compute_instance_id: The instance id for the current node.
+        This is only used in inferring volumes from the cluster_state. This
+        value will be in the attached_to field of the generated
+        ``BlockDeviceVolume`` instances.
+    """
+    node_state = cluster_state.get_node(
+        block_device_deployer.node_uuid,
+        hostname=block_device_deployer.hostname
+    )
+    nonmanifest_datasets = cluster_state.nonmanifest_datasets
+    api = block_device_deployer.block_device_api
+    if volumes is _INFER_VOLUMES_FROM_STATE:
+        if compute_instance_id is None:
+            compute_instance_id = api.compute_instance_id()
+        volumes = _infer_volumes_for_test(
+            block_device_deployer.node_uuid, block_device_deployer.hostname,
+            compute_instance_id, cluster_state)
+    elif volumes is _GET_VOLUMES_FROM_API:
```

Comment:
this case looks like it's used only in one test, so maybe just do the action manually there and simplify this function.